Source/Core/Core/Slippi/SlippiMatchmaking.cpp

```diff
@@ -301,58 +327,58 @@ void SlippiMatchmaking::startMatchmaking()
-	//    would always fetch my VirtualBox IP, which is not correct. I also think perhaps
-	//    it didn't work on Linux/Mac but I haven't tested it.
-	// I left this logic on for now under the assumption that it will help more people than
-	// it will hurt
+	// Determine local IP address. We can attempt to connect to our opponent via
+	// local IP address if we have the same external IP address. The following
+	// scenarios can cause us to have the same external IP address:
+	// - we are connected to the same LAN
+	// - we are connected to the same VPN node
+	// - we are behind the same CGNAT
 	char lanAddr[30] = "";
-
-	char host[256];
-	char *IP;
-	struct hostent *host_entry;
-	int hostname;
-	hostname = gethostname(host, sizeof(host)); // find the host name
-	if (hostname == -1)
+	if (SConfig::GetInstance().m_slippiForceLanIp)
 	{
-		ERROR_LOG(SLIPPI_ONLINE, "[Matchmaking] Error finding LAN address");
+		WARN_LOG(SLIPPI_ONLINE, "[Matchmaking] Overwriting LAN IP sent with configured address");
+		sprintf(lanAddr, "%s:%d", SConfig::GetInstance().m_slippiLanIp.c_str(), m_hostPort);
 	}
 	else
 	{
-		host_entry = gethostbyname(host); // find host information
-		if (host_entry == NULL || host_entry->h_addrtype != AF_INET)
-		{
-			ERROR_LOG(SLIPPI_ONLINE, "[Matchmaking] Error finding LAN host");
-		}
+		enet_uint32 localAddress = getLocalAddress(&addr);
+		if (localAddress != 0)
+			sprintf(lanAddr, "%s:%d", inet_ntoa(*(struct in_addr *)&localAddress), m_hostPort);
 		else
 		{
-			// Fetch the last IP (because that was correct for me, not sure if it will be for all)
-			int i = 0;
-			while (host_entry->h_addr_list[i] != 0)
+			// Fallback: just try the last local IP address listed. They seem to be
+			// listed in decreasing order IE. 192.168.0.100 > 192.168.0.10 > 10.0.0.2
+			char host[256];
+			char *IP;
+			struct hostent *host_entry;
+			int hostname;
+			hostname = gethostname(host, sizeof(host)); // find the host name
+			if (hostname == -1)
 			{
-				IP = inet_ntoa(*((struct in_addr *)host_entry->h_addr_list[i]));
-				WARN_LOG(SLIPPI_ONLINE, "[Matchmaking] IP at idx %d: %s", i, IP);
-				i++;
+				ERROR_LOG(SLIPPI_ONLINE, "[Matchmaking] Error finding LAN address");
+			}
+			else
+			{
+				host_entry = gethostbyname(host); // find host information
+				if (host_entry == NULL || host_entry->h_addrtype != AF_INET)
+				{
+					ERROR_LOG(SLIPPI_ONLINE, "[Matchmaking] Error finding LAN host");
+				}
+				else
+				{
+					// Fetch the last IP (because that was correct for me, not sure if it will be for all)
+					int i = 0;
+					while (host_entry->h_addr_list[i] != 0)
+					{
+						IP = inet_ntoa(*((struct in_addr *)host_entry->h_addr_list[i]));
+						WARN_LOG(SLIPPI_ONLINE, "[Matchmaking] IP at idx %d: %s", i, IP);
+						i++;
+					}
+					sprintf(lanAddr, "%s:%d", IP, m_hostPort);
+				}
 			}
-
-			sprintf(lanAddr, "%s:%d", IP, m_hostPort);
 		}
 	}
-
-	if (SConfig::GetInstance().m_slippiForceLanIp)
```

Comment:
You're confident enough this will work for everyone to remove the override?

Reply:
I moved this block to the top (line 337) to make the priority order more clear (override > new determination mechanism > fallback old determination mechanism)